src/scheduler/parse.c

```diff
@@ -1161,3 +1181,57 @@ preempt_cmp(const void *p1, const void *p2)
 	}
 }
 
+/**
+ * @brief
+ *	       Frees the memory for the corresponding sort_info structure variable given.
+ *
+ * @param[in]	si_type	- type of sort_info structure
+ *
+ * @return void
+ *
+ */
+void
+free_sort_info(enum sort_info_type si_type)
+{
+	struct sort_info *so_info;
+
+	switch (si_type) {
+		case PRIME_SORT:
+			so_info = conf.prime_sort;
+			break;
+		case NON_PRIME_SORT:
+			so_info = conf.non_prime_sort;
+			break;
+		case PRIME_NODE_SORT:
+			so_info = conf.prime_node_sort;
+			break;
+		case NON_PRIME_NODE_SORT:
+			so_info = conf.non_prime_node_sort;
+			break;
+		default:
+			return;
+	}
+
+	if (so_info != NULL) {
+		int i;
+		for (i = 0; i < MAX_SORTS; i++) {
+			if (so_info[i].res_name != NULL) {
+				free(so_info[i].res_name);
+				if (si_type == PRIME_SORT || si_type == PRIME_NODE_SORT) {
+					if (si_type == PRIME_SORT) {
+						if (so_info[i].res_name != conf.non_prime_sort[i].res_name)
+							so_info[i].res_name = NULL;
+						else
+							so_info[i].res_name = conf.non_prime_sort[i].res_name = NULL;
```

Comment:
I'm not sure I understand this.  Wouldn't setting conf.non_prime_sort[i].res_name to NULL leak it?  Could you add a code comment explaining why this is necessary?

Reply:
I have refactored the function free_sort_info() further. Please take a look. 

As there is a possibility that a memory location pointed by res_name can also be referred from another sort_info structures like non_prime_sort we should initialise the corresponding variable in the other sort_info structure to NULL to avoid double free or program crash. And also we need to make sure that free_sort_info() calls should work irrespective of the sequence in which it is called for PRIME_SORT, NON_PRIME_SORT, PRIME_NODE_SORT and NON_PRIME_NODE_SORT.

Example: 
	free_sort_info(PRIME_SORT);
	free_sort_info(NON_PRIME_SORT);
	free_sort_info(PRIME_NODE_SORT);
	free_sort_info(NON_PRIME_NODE_SORT);
OR
      free_sort_info(NON_PRIME_SORT);
      free_sort_info(PRIME_SORT);
      ...
Also I have added comments at the appropriate places in the code. I have made sure that I have run several test suites in valgrind to make sure we do not have any invalid reads/writes or any more leaks after the fix.

Reply:
Would it just be easier to always allocate memory for the name?  It's such a small amount of memory, we're not saving much by reusing it.  It makes this area of the code more confusing to clean up, and adds the possibility for future bugs/double free()s.

Reply:
I have refactored free_sort_info() and also not using the read-only memory at the required place which avoids the need of adding complex logic to the free_sort_info() function. Please look into the latest commit for more details.